azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt

```diff
@@ -173,67 +171,124 @@ class MultiOracleScenarioRunner<
 
     private class ActionGeneratingSystemCursor<AF : ActionFactory, CF : CheckFactory, S : ActionGeneratingSystem<AF, CF>, SF : ActionGeneratingSystemFactory<AF, CF, *, *, *, S>>(
         private val systemFactory: SF,
-        private var systemDefinition: SystemDefinition
+        private val declarations: List<Action>,
     ) {
 
-        private val system = systemFactory.create(systemDefinition)
-
-        fun apply(iteration: SystemIteration) =
-            if (system is MutableSystem<*, *>) {
-                system.applyIteration(iteration)
-                system
-            } else {
-                systemDefinition = systemDefinition.copy(
-                    commands = systemDefinition.commands + iteration.commands,
-                    actionGenerators = iteration.actionGenerators,
-                )
-                systemFactory.create(systemDefinition)
+        private val systemDefinition: SystemDefinition = SystemDefinition(declarations = declarations)
+        private var system = systemFactory.create(systemDefinition)
+        private val unappliedCommands = mutableListOf<Action>()
+        private val commands = mutableListOf<Action>()
+        private val declarationCreators = mutableListOf<(AF) -> Action>()
+        private val commandCreators = mutableListOf<(AF) -> Action>()
+
+        fun generateDeclarations(actionGenerators: List<ActionGenerator>) {
+            checkForUnsupported(actionGenerators)
+            system = systemFactory.create(systemDefinition.copy(
+                declarations = systemDefinition.declarations + declarationCreators.map { it(systemFactory.actionFactory) },
+                actionGenerators = actionGenerators,
+            ))
+            try {
+                declarationCreators.addAll(system.generateActions())
+            } catch (e: LateDetectUnsupportedActionException) {
+                throw ActionGenerationException("Unsupported action generated", e)
+            } catch (e: Exception) {
+                throw ActionGenerationException("Error in declaration action generation", e)
+            } finally {
+                destroy()
+            }
+        }
+
+        private fun checkForUnsupported(actionGenerators: List<ActionGenerator>) {
+            if (actionGenerators.filterIsInstance<UnsupportedActionGenerator>().isNotEmpty()) {
+                throw ActionGenerationException("Unsupported action generator present")
             }
+        }
+
+        fun applyCommands(commands: List<Action>) {
+            system.let {
+                if (it is MutableSystem<*, *>) {
+                    system = systemFactory.create(systemDefinition.copy(
+                        declarations = systemDefinition.declarations + declarationCreators.map { it(systemFactory.actionFactory) },
+                    ))
+                }
+            }
+            unappliedCommands.addAll(commands)
```

Comment:
❓I think I'm missing something here... it seems odd for a function called `applyCommands` to append a list of commands to `unappliedCommands` but do nothing with them.

Reply:
you're mixing the view from outside the box with that inside.. applying the commands from outside stages them to be applied on the next generation cycle.. if there is no further generation cycle then there is no need/desire to create a system with them applied